This project uses the Kaggle Dataset ["When do children learn words?"](https://www.kaggle.com/rtatman/when-do-children-learn-words)

In [84]:
import pandas as pd #data manipulation & analysis  
import numpy as np #math

# Data Initialization 

In [85]:
#import the "main_data" CSV from Data folder
df = pd.read_csv("Data/main_data.csv") 

#import the "Norwegian_CDS_frequency" CSV from Data folder 
freq_df = pd.read_csv("Data/Norwegian_CDS_frequency.csv")

## main_data_df feature & metadata EDA 

In [86]:
df.head()

ID_CDI_I ID_CDI_II           Word_NW          Word_CDI    Translation   AoA  \
0    i_4_1     i_1_1              'au'              'au'         'ouch'  16.0   
1    i_4_2     i_1_2             'bææ'             'bææ'      'baa baa'  15.0   
2    i_4_3     i_1_3  'brrr (bil-lyd)'  'brrr (bil-lyd)'        'vroom'  13.0   
3    i_4_4     i_1_4       'gakk gakk'       'gakk gakk'  'quack quack'  17.0   
4    i_4_5     i_1_5             'grr'             'grr'          'grr'  22.0   

    VSoA        Lex_cat Broad_lex    Freq  CDS_freq  
0   40.0  sound effects  nominals  4366.0       7.0  
1   40.0  sound effects  nominals    18.0       5.0  
2   20.0  sound effects  nominals     NaN      20.0  
3   40.0  sound effects  nominals    16.0       3.0  
4  220.0  sound effects  nominals    78.0       1.0

In [87]:
df.dtypes

ID_CDI_I        object
ID_CDI_II       object
Word_NW         object
Word_CDI        object
Translation     object
AoA            float64
VSoA           float64
Lex_cat         object
Broad_lex       object
Freq           float64
CDS_freq       float64
dtype: object

In [88]:
df.shape

(731, 11)

In [89]:
df.columns

Index(['ID_CDI_I', 'ID_CDI_II', 'Word_NW', 'Word_CDI', 'Translation', 'AoA',
       'VSoA', 'Lex_cat', 'Broad_lex', 'Freq', 'CDS_freq'],
      dtype='object')

In [90]:
df.isna().sum()

ID_CDI_I       341
ID_CDI_II        0
Word_NW          0
Word_CDI         0
Translation      0
AoA             36
VSoA            27
Lex_cat         16
Broad_lex       16
Freq            10
CDS_freq         8
dtype: int64

This dataset contains 731 enteries, 11 columns. There are missing enteries in 7 columns, that will need to be handled before analysis. Reference for columns: 
* IDCDII: Word ID from the Norwegian adaptation of the MacArthur-Bates Communicative Development Inventories, version 1
* IDCDIII: Word ID from the Norwegian adaptation of the MacArthur-Bates Communicative Development Inventories, version 2
* Word_NW: The word in Norwegian
* Word_CDI: The form of the word found in the Norwegian adaptation of the MacArthur-Bates Communicative Development Inventories
* Translation: the English translation of the Norwegian word
* AoA: how old a child generally is was when they this this word, in months (Estimated from the MacArthur-Bates Communicative Development Inventories)
* VSoA: how many other words a child generally knows when they learn this word (rounded up to the nearest 10)
* Lex_cat: the specific part of speech of the word
* Broad_lex: the broad part of speech of the word
* Freq: a measure of how commonly this word occurs in Norwegian
* CDS_Freq: a measure of how commonly this word occurs when a Norwegian adult is talking to a Norwegian child

### DF Data Cleaning 

In [91]:
df.loc[df['ID_CDI_I'].isna()]

ID_CDI_I ID_CDI_II                          Word_NW  \
9        NaN    i_1_10  'oi (uttrykk for overraskelse)'   
20       NaN     i_2_9                        'et esel'   
27       NaN    i_2_16                        'en hane'   
29       NaN    i_2_18                        'ei høne'   
34       NaN    i_2_23                  'en krokodille'   
..       ...       ...                              ...   
726      NaN    i_22_5                           'hvis'   
727      NaN    i_22_6                            'men'   
728      NaN    i_22_7                             'og'   
729      NaN    i_22_8                             'så'   
730      NaN    i_22_9                            'som'   

                            Word_CDI  Translation   AoA   VSoA  \
9    'oi (uttrykk for overraskelse)'         'oh'  20.0   80.0   
20                            'esel'     'donkey'  32.0  560.0   
27                            'hane'    'rooster'  25.0  440.0   
29                            'høne'        'hen'  24.0  280.0   
34                      'krokodille'  'crocodile'  24.0  280.0   
..                               ...          ...   ...    ...   
726                           'hvis'         'if'  35.0  660.0   
727                            'men'        'but'  33.0  600.0   
728                             'og'        'and'  25.0  400.0   
729                             'så'       'then'  31.0  580.0   
730                            'som'       'that'  32.0  600.0   

                Lex_cat     Broad_lex        Freq  CDS_freq  
9         sound effects      nominals      5019.0     437.0  
20         common nouns      nominals      1292.0       9.0  
27         common nouns      nominals      1508.0       1.0  
29         common nouns      nominals      4401.0       5.0  
34         common nouns      nominals      1190.0      15.0  
..                  ...           ...         ...       ...  
726  closed-class items  closed-class    498529.0      86.0  
727  closed-class items  closed-class   3015440.0     444.0  
728  closed-class items  closed-class  16079937.0    1074.0  
729  closed-class items  closed-class   2141716.0     900.0  
730  closed-class items  closed-class   5486715.0     402.0  

[341 rows x 11 columns]

341 rows in this column are NaN, there is a second ID column ID_CDI_II which has no NaN values, to handle these NaN values I am going to drop the first column

In [92]:
df = df.drop(columns =['ID_CDI_I'])

In [93]:
df.head()

ID_CDI_II           Word_NW          Word_CDI    Translation   AoA   VSoA  \
0     i_1_1              'au'              'au'         'ouch'  16.0   40.0   
1     i_1_2             'bææ'             'bææ'      'baa baa'  15.0   40.0   
2     i_1_3  'brrr (bil-lyd)'  'brrr (bil-lyd)'        'vroom'  13.0   20.0   
3     i_1_4       'gakk gakk'       'gakk gakk'  'quack quack'  17.0   40.0   
4     i_1_5             'grr'             'grr'          'grr'  22.0  220.0   

         Lex_cat Broad_lex    Freq  CDS_freq  
0  sound effects  nominals  4366.0       7.0  
1  sound effects  nominals    18.0       5.0  
2  sound effects  nominals     NaN      20.0  
3  sound effects  nominals    16.0       3.0  
4  sound effects  nominals    78.0       1.0

In [100]:
df.loc[df['VSoA'] == 660.0]

ID_CDI_II            Word_NW        Word_CDI              Translation  \
25     i_2_14           'ei gås'           'gås'                  'goose'   
45     i_2_34      'et reinsdyr'      'reinsdyr'               'reindeer'   
54     i_2_43          'en valp'          'valp'                  'puppy'   
80     i_4_11         'et kritt'         'kritt'                  'chalk'   
126    i_5_40          'nugatti'       'nugatti'                'Nugatti'   
224    i_8_12         'en gåbil'         'gåbil'                'toy car'   
273    i_9_11        'gyngestol'     'gyngestol'          'rocking chair'   
291    i_9_29  'en tørketrommel'  'tørketrommel'                  'dryer'   
343   i_11_16          'en park'          'park'                   'park'   
376   i_12_27       'en oldefar'       'oldefar'      'great-grandfather'   
598   i_16_16           'ei tid'           'tid'                   'time'   
602    i_17_4              'dem'           'dem'                   'them'   
606    i_17_8            'deres'         'deres'  'your(s) (pl)/their(s)'   
615   i_17_17            'henne'         'henne'                    'her'   
618   i_17_20        'hverandre'     'hverandre'             'each other'   
622   i_17_24         'meg selv'      'meg selv'                 'myself'   
632    i_18_3          'hvilken'       'hvilken'                  'which'   
634    i_18_5          'hvordan'       'hvordan'                    'how'   
658   i_19_22              'mot'           'mot'        'towards/against'   
674   i_19_38          'utenfor'       'utenfor'                'outside'   
717   i_21_18         'å skulle'        'skulle'                  'would'   
726    i_22_5             'hvis'          'hvis'                     'if'   

      AoA   VSoA             Lex_cat     Broad_lex       Freq  CDS_freq  
25   36.0  660.0        common nouns      nominals     1491.0       2.0  
45   36.0  660.0        common nouns      nominals     1317.0       3.0  
54   34.0  660.0        common nouns      nominals    26852.0       2.0  
80   35.0  660.0        common nouns      nominals      550.0       1.0  
126  36.0  660.0        common nouns      nominals      209.0       1.0  
224  36.0  660.0        common nouns      nominals        3.0       1.0  
273  35.0  660.0        common nouns      nominals      188.0       1.0  
291  35.0  660.0        common nouns      nominals      767.0       4.0  
343  34.0  660.0        places to go      nominals    14131.0      19.0  
376  33.0  660.0              people      nominals     1234.0       1.0  
598  35.0  660.0                 NaN           NaN   644085.0      11.0  
602  36.0  660.0  closed-class items  closed-class   462725.0      39.0  
606  34.0  660.0  closed-class items  closed-class   201042.0       3.0  
615  35.0  660.0  closed-class items  closed-class   170695.0      12.0  
618  36.0  660.0  closed-class items  closed-class   101055.0      15.0  
622  35.0  660.0  closed-class items  closed-class    25540.0       1.0  
632  35.0  660.0  closed-class items  closed-class   240992.0      70.0  
634  35.0  660.0  closed-class items  closed-class   435967.0      33.0  
658  35.0  660.0  closed-class items  closed-class   849778.0       5.0  
674  35.0  660.0  closed-class items  closed-class   113560.0       7.0  
717  34.0  660.0  closed-class items  closed-class  2295140.0    1152.0  
726  35.0  660.0  closed-class items  closed-class   498529.0      86.0

There are 36 values in the AoA column that are NaN. There is some correlation between this value and the value found in the VSoA column. A VSoA value of 660.0 equals as AoA value of between 34.0 - 36.0. As such I am going to fill in missing AoA values with a known VSoA value of 660.0 to the mean of the known AoA range = 35.0 

In [113]:
#replace the nan values in AoA based on VSoA
df['AoA'] = np.where((df['AoA'].isna()) & ((df['VSoA'] == 660.0)), 35.0, df['AoA']) 
df['AoA'] = np.where((df['AoA'].isna()) & ((df['VSoA'] == 680.0)), 35.0, df['AoA'])

In [114]:
#sanity check
df.loc[df['AoA'].isna()]

ID_CDI_II            Word_NW           Word_CDI             Translation  \
93      i_5_6         'en bønne'           'bønner'                 'beans'   
104    i_5_18  'frokostblanding'  'frokostblanding'                'cereal'   
137    i_5_51      'ristet brød'      'ristet brød'                 'toast'   
185     i_7_1         'en ankel'            'ankel'                 'ankle'   
232    i_8_20      'kjøkkenrull'      'kjøkkenrull'  'roll of paper towels'   
237    i_8_25        'ei krukke'           'krukke'                   'jar'   
265     i_9_3       'en balkong'          'balkong'               'balcony'   
298    i_10_2          'bakgård'          'bakgård'              'backyard'   
332    i_11_5  'en campingplass'     'campingplass'          'camping site'   
341   i_11_14           'landet'           'landet'       'the countryside'   
359   i_12_10           'cowboy'           'cowboy'                'cowboy'   
366   i_12_17   'ei helsesøster'      'helsesøster'                 'nurse'   
436   i_14_24           'å hate'             'hate'                  'hate'   
441   i_14_29  'å holde øye med'    'holde øye med'                 'watch'   
533   i_15_14            'frekk'            'frekk'                'cheeky'   
562   i_15_43           'skjønn'           'skjønn'             'beautiful'   
566   i_15_47           'søvnig'           'søvnig'                'sleepy'   
585    i_16_3      'ettermiddag'      'ettermiddag'             'afternoon'   
588    i_16_6        'formiddag'        'formiddag'               'morning'   
601    i_17_3         'deg selv'         'deg selv'              'yourself'   
612   i_17_14              'ham'              'ham'                   'him'   
620   i_17_22              'man'              'man'                   'one'   
626   i_17_28         'seg selv'         'seg selv'       'himself/herself'   
653   i_19_17     'i stedet for'     'i stedet for'            'instead of'   
688   i_20_11            'hvert'            'hvert'                  'each'   
701    i_21_2              'bør'              'bør'                'should'   
702    i_21_3          'å burde'            'burde'                'should'   

     AoA  VSoA             Lex_cat     Broad_lex       Freq  CDS_freq  
93   NaN   NaN        common nouns      nominals     3285.0       1.0  
104  NaN   NaN        common nouns      nominals      686.0       1.0  
137  NaN   NaN        common nouns      nominals      108.0       1.0  
185  NaN   NaN        common nouns      nominals     3147.0       1.0  
232  NaN   NaN        common nouns      nominals        9.0       1.0  
237  NaN   NaN        common nouns      nominals     1312.0       1.0  
265  NaN   NaN        common nouns      nominals     3835.0       2.0  
298  NaN   NaN        places to go      nominals     2581.0       1.0  
332  NaN   NaN        places to go      nominals     3425.0       1.0  
341  NaN   NaN        places to go      nominals        NaN       NaN  
359  NaN   NaN              people      nominals     1734.0       8.0  
366  NaN   NaN              people      nominals     4949.0       1.0  
436  NaN   NaN        action words    predicates    25992.0       1.0  
441  NaN   NaN        action words    predicates      818.0       1.0  
533  NaN   NaN   descriptive words    predicates     8977.0       2.0  
562  NaN   NaN   descriptive words    predicates    12524.0       1.0  
566  NaN   NaN   descriptive words    predicates      921.0       1.0  
585  NaN   NaN                 NaN           NaN    29317.0       3.0  
588  NaN   NaN                 NaN           NaN    14515.0       2.0  
601  NaN   NaN  closed-class items  closed-class    33409.0       1.0  
612  NaN   NaN  closed-class items  closed-class   172668.0       3.0  
620  NaN   NaN  closed-class items  closed-class  1383204.0      55.0  
626  NaN   NaN  closed-class items  closed-class   130150.0       1.0  
653  NaN   NaN  closed-class items  closed-class    32260.0    

The final NaN values in AoA are without reference value in the VSoA, so I will drop these columns. 

In [115]:
df = df.dropna(subset=['AoA'])

In [117]:
df.loc[df['Lex_cat'].isna()]

ID_CDI_II     Word_NW    Word_CDI   Translation   AoA   VSoA Lex_cat  \
583    i_16_1    'en dag'       'dag'         'day'  26.0  360.0     NaN   
584    i_16_2     'etter'     'etter'       'after'  31.0  580.0     NaN   
586    i_16_4   'etterpå'   'etterpå'  'afterwords'  25.0  360.0     NaN   
587    i_16_5       'før'       'før'      'before'  30.0  580.0     NaN   
589    i_16_7     'i dag'     'i dag'       'today'  27.0  440.0     NaN   
590    i_16_8     'i går'     'i går'   'yesterday'  30.0  580.0     NaN   
591    i_16_9   'i kveld'   'i kveld'     'tonight'  32.0  620.0     NaN   
592   i_16_10  'i morgen'  'i morgen'    'tomorrow'  28.0  460.0     NaN   
593   i_16_11  'en kveld'     'kveld'     'evening'  28.0  520.0     NaN   
594   i_16_12    'morgen'    'morgen'     'morning'  28.0  520.0     NaN   
595   i_16_13        'nå'        'nå'         'now'  25.0  400.0     NaN   
596   i_16_14   'ei natt'      'natt'       'night'  24.0  260.0     NaN   
597   i_16_15    'senere'    'senere'       'later'  30.0  580.0     NaN   
598   i_16_16    'ei tid'       'tid'        'time'  35.0  660.0     NaN   

    Broad_lex       Freq  CDS_freq  
583       NaN  1074722.0      62.0  
584       NaN  1137455.0      23.0  
586       NaN    36076.0      29.0  
587       NaN   598533.0      41.0  
589       NaN   394236.0       2.0  
590       NaN    83408.0       7.0  
591       NaN     2709.0       2.0  
592       NaN    37720.0       7.0  
593       NaN   123627.0       2.0  
594       NaN    46248.0       2.0  
595       NaN  1120297.0     461.0  
596       NaN    73005.0       4.0  
597       NaN   125096.0       1.0  
598       NaN   644085.0      11.0

The Lex_cat NaN words corespont to the Broad_lex NaN values, removing these columns will solve for Nans in both. 

In [120]:
df = df.dropna(subset=['Lex_cat'])

In [121]:
df.loc[df['Freq'].isna()]

ID_CDI_II              Word_NW             Word_CDI          Translation  \
2       i_1_3     'brrr (bil-lyd)'     'brrr (bil-lyd)'              'vroom'   
12      i_2_1        '(meite)mark'        '(meite)mark'        '(earth)worm'   
92      i_5_5              'bolle'             'boller'               'buns'   
221    i_8_10                'fat'                'fat'      '(large) plate'   
352    i_12_3       'barnets navn'       'barnets navn'   'child's own name'   
353    i_12_4  'barnevaktens navn'  'barnevaktens navn'  'babysitter's name'   
375   i_12_26   'navn på kjæledyr'   'navn på kjæledyr'         'pet's name'   
443   i_14_31            'å huske'              'huske'              'swing'   
455   i_14_43            'å koste'              'koste'          cost/sweep'   

      AoA   VSoA        Lex_cat   Broad_lex  Freq  CDS_freq  
2    13.0   20.0  sound effects    nominals   NaN      20.0  
12   28.0  460.0   common nouns    nominals   NaN       NaN  
92   20.0  120.0   common nouns    nominals   NaN       NaN  
221  26.0  400.0   common nouns    nominals   NaN       NaN  
352  20.0  100.0         people    nominals   NaN     317.0  
353  25.0  360.0         people    nominals   NaN       NaN  
375  25.0  240.0         people    nominals   NaN       NaN  
443  25.0  300.0   action words  predicates   NaN       NaN  
455  26.0  400.0   action words  predicates   NaN       NaN

The Freq Nan & CDS_freq do not provide a whole picture of the word so I will drop these nans as well. 

In [122]:
df = df.dropna(subset=['Freq', 'CDS_freq'])

In [125]:
#confirm all nans removed 
df.isna().sum()

ID_CDI_II      0
Word_NW        0
Word_CDI       0
Translation    0
AoA            0
VSoA           0
Lex_cat        0
Broad_lex      0
Freq           0
CDS_freq       0
dtype: int64

## Norwegian_CDS_frequency_df feature & metadata EDA

In [40]:
freq_df.head()

Word_CDI    Translation  Freq_NoWaC  Freq_CDS
0              'au'         'ouch'      4366.0       7.0
1             'bææ'      'baa baa'        18.0       5.0
2  'brrr (bil-lyd)'        'vroom'         NaN      20.0
3       'gakk gakk'  'quack quack'        16.0       3.0
4             'grr'          'grr'        78.0       1.0

In [41]:
freq_df.dtypes

Word_CDI        object
Translation     object
Freq_NoWaC     float64
Freq_CDS       float64
dtype: object

In [42]:
freq_df.shape

(731, 4)

In [43]:
freq_df.columns

Index(['Word_CDI', 'Translation', 'Freq_NoWaC', 'Freq_CDS'], dtype='object')

In [44]:
freq_df.isna().sum()

Word_CDI        0
Translation     0
Freq_NoWaC     10
Freq_CDS        8
dtype: int64

This dataset contains 731 entries and 4 columns. 2 columns have null values that will have to be resolved before analysis. Reference for columns: 
* Word_CDI: The word from, as found in the Norwegian adaptation of the MacArthur-Bates Communicative Development Inventories
* Translation: The English translation of the Norwegian word
* Freq_NoWaC: How often this word is used on the internet
* Freq_CDS: How often this word is used when talking to children (based on two Norwegian CHILDES corpora)

### DF Data Cleaning 

In [126]:
freq_df.loc[freq_df['Freq_NoWaC'].isna()]

Word_CDI          Translation  Freq_NoWaC  Freq_CDS
2       'brrr (bil-lyd)'              'vroom'         NaN      20.0
12         '(meite)mark'        '(earth)worm'         NaN       NaN
92              'boller'               'buns'         NaN       NaN
221                'fat'      '(large) plate'         NaN       NaN
341             'landet'    'the countryside'         NaN       NaN
352       'barnets navn'   'child's own name'         NaN     317.0
353  'barnevaktens navn'  'babysitter's name'         NaN       NaN
375   'navn på kjæledyr'         'pet's name'         NaN       NaN
443              'huske'              'swing'         NaN       NaN
455              'koste'          cost/sweep'         NaN       NaN

This set of NaN values only accounts for 10 words (8 of which also have NaN values for Freq_CDS fields) to handle these I will drop them from the dataframe. 

In [127]:
freq_df = freq_df.dropna()

In [128]:
#sanity check 
freq_df.isna().sum()

Word_CDI       0
Translation    0
Freq_NoWaC     0
Freq_CDS       0
dtype: int64

# Exploratory Analysis 

In [129]:
#join dataframes on the Word CDI 
pd.concatenate

ID_CDI_II      Word_NW     Word_CDI          Translation   AoA   VSoA  \
0     i_1_1         'au'         'au'               'ouch'  16.0   40.0   
1     i_1_2        'bææ'        'bææ'            'baa baa'  15.0   40.0   
3     i_1_4  'gakk gakk'  'gakk gakk'        'quack quack'  17.0   40.0   
4     i_1_5        'grr'        'grr'                'grr'  22.0  220.0   
5     i_1_6   'kykeliky'   'kykeliky'  'cock-a-doodle-doo'  22.0  220.0   

         Lex_cat Broad_lex    Freq  CDS_freq  
0  sound effects  nominals  4366.0       7.0  
1  sound effects  nominals    18.0       5.0  
3  sound effects  nominals    16.0       3.0  
4  sound effects  nominals    78.0       1.0  
5  sound effects  nominals     8.0       2.0

In [131]:
df['ID_CDI_II'].values

array(['i_1_1', 'i_1_2', 'i_1_4', 'i_1_5', 'i_1_6', 'i_1_7', 'i_1_8',
       'i_1_9', 'i_1_10', 'i_1_11', 'i_1_12', 'i_2_2', 'i_2_3', 'i_2_4',
       'i_2_5', 'i_2_6', 'i_2_7', 'i_2_8', 'i_2_9', 'i_2_10', 'i_2_11',
       'i_2_12', 'i_2_13', 'i_2_14', 'i_2_15', 'i_2_16', 'i_2_17',
       'i_2_18', 'i_2_19', 'i_2_20', 'i_2_21', 'i_2_22', 'i_2_23',
       'i_2_24', 'i_2_25', 'i_2_26', 'i_2_27', 'i_2_28', 'i_2_29',
       'i_2_30', 'i_2_31', 'i_2_32', 'i_2_33', 'i_2_34', 'i_2_35',
       'i_2_36', 'i_2_37', 'i_2_38', 'i_2_39', 'i_2_40', 'i_2_41',
       'i_2_42', 'i_2_43', 'i_2_44', 'i_3_1', 'i_3_2', 'i_3_3', 'i_3_4',
       'i_3_5', 'i_3_6', 'i_3_7', 'i_3_8', 'i_3_9', 'i_3_10', 'i_3_11',
       'i_3_12', 'i_3_13', 'i_3_14', 'i_4_1', 'i_4_2', 'i_4_3', 'i_4_4',
       'i_4_5', 'i_4_6', 'i_4_7', 'i_4_8', 'i_4_9', 'i_4_10', 'i_4_11',
       'i_4_12', 'i_4_13', 'i_4_14', 'i_4_15', 'i_4_16', 'i_4_17',
       'i_4_18', 'i_5_1', 'i_5_2', 'i_5_3', 'i_5_4', 'i_5_7', 'i_5_8',
       'i_5_9', 'i_5_1